### PR TITLE
ch4: default MPIDI_CH4_MAX_VCIS to 64 when per-vci is on

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Currently, user have to configure with both --with-thread-cs=per-vci and
--with-ch4-max-vcis, as well as set MPIR_CVAR_CH4_NUM_VCIS to enable
mutiple vcis. This is a bit cumbersome to remember and to explain. Let's
default `--with-ch4-max-vcis` to 64 -- the maximum number of request
pools -- to make it more convenient. It only adds 100K or so global
memory, which most applications won't be concerned with.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom --with-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Review by raffenet — APPROVED** [review]

File: src/mpid/ch4/subconfigure.m4 [hunk 1]
```diff
@@ -333,13 +333,23 @@ fi
...
+else
+    if test $thread_granularity != MPICH_THREAD_GRANULARITY__VCI ; then
+        AC_MSG_ERROR(Option --with-ch4-max-vcis requires --enable-thread-cs=per-vci)
+    fi
 fi
-if test $with_ch4_max_vcis -gt 1 -a $thread_granularity != MPICH_THREAD_GRANULARITY__VCI ; then
-    AC_MSG_ERROR(CH4_MAX_VCIS greater than 1 requires --enable-thread-cs=per-vci)
+
+if test $with_ch4_max_vcis -lt 1 -o $with_ch4_max_vcis -gt 64; then
+   AC_MSG_ERROR(Number of VCIs must be between 1 and 64)
```
Comment (sagarth) [inline src/mpid/ch4/subconfigure.m4:352]: Did you think of any specific reason to cap the number of vcis at 64?

Comment (hzhou) [reply src/mpid/ch4/subconfigure.m4:352]: Yes, the request handle currently use 6 bits to code pools, thus capped at 64.

Comment (sagarth) [reply src/mpid/ch4/subconfigure.m4:352]: I see. I think it would be good to document it somewhere (if not done already). Maybe as a code comment here?

**Comment by hzhou** [comment]
test:mpich/custom --with-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Review by sagarth — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/custom --enable-thread-cs=per-vci
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

**Pull request merged**